File: NC/reorder-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution
{
public:
    ListNode *reverseList(ListNode *head)
    {
        if (!head || head->next == NULL)
        {
            return head;
        }

        ListNode *next = head->next;
        head->next = NULL;
        ListNode *newHead = reverseList(next);
        next->next = head;
        return newHead;
    }

    void reorderList(ListNode *head)
    {

        ListNode *curr = head;
        int count = 0;

        while (curr)
        {
            count++;
            curr = curr->next;
        }

        count = (count + 1) / 2;

        curr = head;

        for (int i = 1; i < count; i++)
        {
            curr = curr->next;
        }

        ListNode *newList = curr->next;

        curr->next = NULL;

        newList = reverseList(newList);
        curr = head;

        while (curr && newList)
        {
            ListNode *currNext = curr->next, *newListNext = newList->next;
            curr->next = newList;
            newList->next = currNext;
            curr = currNext;
            newList = newListNext;
        }
    }
};

File: NC/java/ReorderLinkedList.java
package NC.java;

public class ReorderLinkedList {
    private ListNode reverseList(ListNode node) {
        if (node == null || node.next == null) {
            return node;
        }

        ListNode next = node.next;
        node.next = null;
        ListNode newHead = reverseList(next);
        next.next = node;

        return newHead;

    }

    public void reorderList(ListNode head) {
        if (head == null) {
            return;
        }
        ListNode curr = head, temp1, temp2;

        int listLength = 1;
        while (curr.next != null) {
            listLength++;
            curr = curr.next;
        }

        if (listLength == 1) {
            return;
        }

        int breakLength = listLength / 2;

        curr = head;

        for (int i = 0; i < breakLength; i++) {
            curr = curr.next;
        }

        ListNode mid = curr.next;
        curr.next = null;

        mid = reverseList(mid);

        curr = head;

        while (mid != null) {

            temp1 = curr.next;
            temp2 = mid.next;

            curr.next = mid;
            mid.next = temp1;
            curr = temp1;
            mid = temp2;
        }

    }
}
